clients/src/main/java/org/apache/kafka/clients/admin/DescribeClusterResults.java

```diff
@@ -28,16 +28,38 @@
  */
 @InterfaceStability.Unstable
 public class DescribeClusterResults {
-    private final KafkaFuture<Collection<Node>> future;
+    private final KafkaFuture<Collection<Node>> nodesFuture;
+    private final KafkaFuture<Node> controllerFuture;
+    private final KafkaFuture<String> clusterIdFuture;
 
-    DescribeClusterResults(KafkaFuture<Collection<Node>> future) {
-        this.future = future;
+    DescribeClusterResults(KafkaFuture<Collection<Node>> nodesFuture,
+                           KafkaFuture<Node> controllerFuture,
+                           KafkaFuture<String> clusterIdFuture) {
+        this.nodesFuture = nodesFuture;
+        this.controllerFuture = controllerFuture;
+        this.clusterIdFuture = clusterIdFuture;
     }
 
     /**
      * Returns a future which yields a collection of nodes.
      */
     public KafkaFuture<Collection<Node>> nodes() {
-        return future;
+        return nodesFuture;
+    }
+
+    /**
+     * Returns a future which yields the current controller id.
+     * Note that this may yield null, if the controller ID is not yet known.
+     */
+    public KafkaFuture<Node> controller() {
+        return controllerFuture;
+    }
+
+    /**
+     * Returns a future which yields the current cluster Id.
+     * Note that this may yield null, if the cluster version is too old.
+     */
+    public KafkaFuture<String> clusterIdFuture() {
```

Comment:
Good call, it should be clusterId, not clusterIdFuture.

We want to have separate fields because it's cleaner (having to pick things out of a giant struct is annoying) and also allows us to have one future throw an error while the other futures succeed in the future if we want.